### PR TITLE
Add UUID support

### LINKED ISSUE #2065: Implement UUID data type

### PR DESCRIPTION
**Pull request by papafe**
## Description
Add support for `UUID` data type. 
In order to use this, the object property must be declared of type `Guid`:

```csharp
public MyObject : RealmObject
{
    public Guid Id {get; set;}

    .... 
}
```
The new data type:
- Can be used as a primary key
- Can be used in queries: equal, not equal
- Is indexable

Fixes #2065 

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm/Native/PrimitiveValue.cs [hunk 1]
```diff
@@ -140,6 +143,24 @@ public static PrimitiveValue ObjectId(ObjectId value)
...
 
+        public static PrimitiveValue Guid(Guid value)
+        {
+            var result = new PrimitiveValue
+            {
+                Type = RealmValueType.Guid,
+            };
+
+            var guidBytes = value.ToByteArray();
+            for (int i = 0; i < 16; i++)
```
Comment (nirinchev) [inline Realm/Realm/Native/PrimitiveValue.cs:154]:
```suggestion
            for (var i = 0; i < 16; i++)
```

File: Tests/Realm.Tests/Database/TestObjects.cs [hunk 2]
```diff
@@ -336,7 +340,14 @@ public class PrimaryKeyObjectIdObject : RealmObject
     {
         [PrimaryKey]
         [MapTo("_id")]
-        public ObjectId StringProperty { get; set; }
```
Comment (nirinchev) [inline Tests/Realm.Tests/Database/TestObjects.cs:339]: 🙈

File: Tests/Weaver/AssemblyToProcess/TestObjects.cs [hunk 3]
```diff
@@ -53,6 +53,9 @@ public class AllTypesObject : RealmObject
 
         public ObjectId ObjectIdProperty { get; set; }
 
+        public Guid GuidProperty { get; set; }
+
+
```
Comment (nirinchev) [inline Tests/Weaver/AssemblyToProcess/TestObjects.cs:58]:
```suggestion

```

File: Tests/Weaver/AssemblyToProcess/TestObjects.cs [hunk 4]
```diff
@@ -92,6 +95,9 @@ public class AllTypesObject : RealmObject
         public RealmInteger<long>? NullableInt64CounterProperty { get; set; }
 
         public ObjectId? NullableObjectIdProperty { get; set; }
+
+        public Guid? NullableGuidProperty { get; set; }
+
```
Comment (nirinchev) [inline Tests/Weaver/AssemblyToProcess/TestObjects.cs:100]:
```suggestion
```

File: wrappers/src/marshalling.hpp [hunk 5]
```diff
@@ -171,6 +171,25 @@ static inline ObjectId from_capi(realm_object_id_t oid)
     return ObjectId(std::move(bytes));
 }
 
+static inline realm_uuid_t to_capi(UUID uuid)
+{
+    auto bytes = uuid.to_bytes();
```
Comment (nirinchev) [inline wrappers/src/marshalling.hpp:176]: This should probably be `auto&` to avoid copying?

**Review by nirinchev — APPROVED** [review]
Pretty good! Need a changelog entry + a few nits, but otherwise seems to be ready to merge!

**Comment by nirinchev** [comment]
Oh, just noticed - can we add a test for `[Indexed]` applied on a Guid property?

**Comment by bmunkholm** [comment]
This should likely not be a "draft" anymore if it's ready for review and merging :-)?

File: Tests/Weaver/AssemblyToProcess/TestObjects.cs (hunk 4 above)
Comment (papafe) [reply Tests/Weaver/AssemblyToProcess/TestObjects.cs:98]: I suppose you wanted to suggest to remove the line after `NullableGuidProperty` and not before. I will do it in a commit :)

**Comment by papafe** [comment]
Added corrections for Weaver tests, added data type to the list of indexable ones and added it to the tests.

**Comment by papafe** [comment]
Slightly modified the copy mechanism according to comment

**Review by LaPeste — APPROVED** [review]

File: Realm/Realm/Schema/PropertyType.cs [hunk 6]
```diff
@@ -85,6 +85,11 @@ public enum PropertyType : ushort
         /// </summary>
         Decimal = 11,
 
+        /// <summary>
+        /// 128 bit UUID property.
+        /// </summary>
+        Guid = 12,
```
Comment (cmelchior) [inline Realm/Realm/Schema/PropertyType.cs:91]: Does this also mean that Lists of Guids are supported? At least I could not find any tests for it?

Comment (nirinchev) [reply Realm/Realm/Schema/PropertyType.cs:91]: They are, yes, but you're right the PR is missing tests for those - completely forgot about it 😅

**Comment by papafe** [comment]
Added support and tests for list of Guids

**Review by nirinchev — COMMENTED** [review]
@papafe I feel bad for the ton of conflicts, so I'll get this rebased and merged tomorrow.

**Pull request merged**